File: cpp/tests/test_utils.cpp
#include "utils.hpp"
#include <cmath>
#include <gtest/gtest.h>
#include <opencv2/opencv.hpp>

namespace rfdetr {

TEST(UtilsTest, SigmoidTest) {
  cv::Mat input = (cv::Mat_<float>(1, 3) << -100.0f, 0.0f, 100.0f);
  cv::Mat output;
  sigmoid(input, output);

  EXPECT_NEAR(output.at<float>(0, 0), 0.0f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 1), 0.5f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 2), 1.0f, 1e-6);
}

TEST(UtilsTest, BoxConversionTest) {
  // Input: cx, cy, w, h
  cv::Mat input = (cv::Mat_<float>(1, 4) << 0.5f, 0.5f, 0.2f, 0.4f);
  cv::Mat output;
  boxCxcywhToXyxyn(input, output);

  // Expected: xmin = 0.5 - 0.1 = 0.4
  //           ymin = 0.5 - 0.2 = 0.3
  //           xmax = 0.5 + 0.1 = 0.6
  //           ymax = 0.5 + 0.2 = 0.7
  EXPECT_NEAR(output.at<float>(0, 0), 0.4f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 1), 0.3f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 2), 0.6f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 3), 0.7f, 1e-6);
}

TEST(UtilsTest, BoxConversionToXywhTest) {
  // Input: cx, cy, w, h
  cv::Mat input = (cv::Mat_<float>(1, 4) << 0.5f, 0.5f, 0.2f, 0.4f);
  cv::Mat output;
  boxCxcywhToXywh(input, output);

  // Expected: x_left = 0.5 - 0.1 = 0.4
  //           y_top = 0.5 - 0.2 = 0.3
  //           w = 0.2
  //           h = 0.4
  EXPECT_NEAR(output.at<float>(0, 0), 0.4f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 1), 0.3f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 2), 0.2f, 1e-6);
  EXPECT_NEAR(output.at<float>(0, 3), 0.4f, 1e-6);
}

} // namespace rfdetr

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: python/tests/test_utils.py
import numpy as np
import pytest
import os
import sys

# Add project root to path
REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "../.."))
PYTHON_DIR = os.path.join(REPO_ROOT, "python")
if PYTHON_DIR not in sys.path:
    sys.path.insert(0, PYTHON_DIR)

from modules.utils import sigmoid, box_cxcywh_to_xyxyn

def test_sigmoid():
    x = np.array([-100.0, 0.0, 100.0])
    expected = 1 / (1 + np.exp(-x))
    output = sigmoid(x)
    
    np.testing.assert_allclose(output, expected, atol=1e-6)
    assert output[0] < 0.01
    assert abs(output[1] - 0.5) < 1e-6
    assert output[2] > 0.99

def test_box_cxcywh_to_xyxyn():
    # Input: cx, cy, w, h
    x = np.array([[0.5, 0.5, 0.2, 0.4]])
    
    # Expected: xmin = 0.5 - 0.1 = 0.4
    #           ymin = 0.5 - 0.2 = 0.3
    #           xmax = 0.5 + 0.1 = 0.6
    #           ymax = 0.5 + 0.2 = 0.7
    expected = np.array([[0.4, 0.3, 0.6, 0.7]])
    output = box_cxcywh_to_xyxyn(x)
    
    np.testing.assert_allclose(output, expected, atol=1e-6)
